Sources/_CryptoExtras/MLDSA/MLDSA_boring.swift

```diff
@@ -0,0 +1,321 @@
+//
+// Copyright (c) 2024 Apple Inc. and the SwiftCrypto project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.md for the list of SwiftCrypto project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Crypto
+import Foundation
+
+@_implementationOnly import CCryptoBoringSSL
+@_implementationOnly import CCryptoBoringSSLShims
+
+/// A lattice-based digital signature algorithm that provides security against quantum computing attacks.
+public enum MLDSA {}
+
+extension MLDSA {
+    /// A ML-DSA-65 private key.
+    public struct PrivateKey: Sendable {
+        private var backing: Backing
+
+        /// Initialize a ML-DSA-65 private key from a random seed.
+        public init() throws {
+            self.backing = try Backing()
+        }
+
+        /// Initialize a ML-DSA-65 private key from a seed.
+        /// 
+        /// The seed must be at least 32 bytes long.
+        /// Any additional bytes in the seed are ignored.
+        /// 
+        /// - Parameter seed: The seed to use to generate the private key.
+        /// 
+        /// - Throws: `CryptoKitError.incorrectKeySize` if the seed is not at least 32 bytes long.
+        public init(seed: some DataProtocol) throws {
+            self.backing = try Backing(seed: seed)
+        }
+
+        /// The seed from which this private key was generated.
+        public var seed: Data {
+            self.backing.seed
+        }
+
+        /// The public key associated with this private key.
+        public var publicKey: PublicKey {
+            self.backing.publicKey
+        }
+
+        /// Generate a signature for the given data.
+        /// 
+        /// - Parameters: 
+        ///   - data: The message to sign.
+        ///   - context: The context to use for the signature.
+        /// 
+        /// - Returns: The signature of the message.
+        public func signature<D: DataProtocol>(for data: D, context: D? = nil) throws -> Signature {
+            try self.backing.signature(for: data, context: context)
+        }
+
+        /// The size of the private key in bytes.
+        static let bytesCount = Backing.bytesCount
+
+        fileprivate final class Backing {
+            private let pointer: UnsafeMutablePointer<MLDSA65_private_key>
+            private let seedPointer: UnsafeMutablePointer<UInt8>
+
+            func withUnsafePointer<T>(_ body: (UnsafePointer<MLDSA65_private_key>) throws -> T) rethrows -> T {
+                try body(self.pointer)
+            }
+
+            /// Initialize a ML-DSA-65 private key from a random seed.
+            init() throws {
+                self.pointer = UnsafeMutablePointer<MLDSA65_private_key>.allocate(capacity: 1)
+                self.seedPointer = UnsafeMutablePointer<UInt8>.allocate(capacity: MLDSA.seedSizeInBytes)
```

Comment:
I think we can make this a touch more efficient. We can store both of these fields in their useful formats. For the private key, we can just have a `var key: MLDSA65_private_key`, and then derive pointers as needed.

For the seed, we can use a `withUnsafeTemporaryAllocation` and then directly create a `Data` from it. `Data` is the format we need elsewhere anyway.

Reply:
This means that we don't need a `Backing` class?

Reply:
Technically it does, but we should keep it to avoid us abusing the power of `@_implementationOnly`.